File: profile/internal/delivery/event/user_registered_listener.go
package event

import (
	"context"
	"encoding/json"
	"time"

	"github.com/nats-io/nats.go"
	"github.com/nats-io/nats.go/jetstream"

	"microservices/profile/internal/config"
	"microservices/profile/internal/domain/service"
	"microservices/profile/pkg/logger"
)

const (
	StreamName   = "USERS"
	ConsumerName = "profile-service-user-consumer"
	Subject      = "user.registered"
)

type UserRegisteredEvent struct {
	UserID       string `json:"userId"`
	Email        string `json:"email"`
	FullName     string `json:"fullName"`
	RegisteredAt string `json:"registeredAt"`
}

type UserRegisteredListener struct {
	cfg            *config.Config
	profileService *service.ProfileService
	conn           *nats.Conn
	js             jetstream.JetStream
	consumer       jetstream.Consumer
	stopCh         chan struct{}
}

func NewUserRegisteredListener(cfg *config.Config, profileService *service.ProfileService) *UserRegisteredListener {
	return &UserRegisteredListener{
		cfg:            cfg,
		profileService: profileService,
		stopCh:         make(chan struct{}),
	}
}

func (l *UserRegisteredListener) Start(ctx context.Context) error {
	// Connect to NATS
	conn, err := nats.Connect(l.cfg.NATS.URL,
		nats.RetryOnFailedConnect(true),
		nats.MaxReconnects(10),
		nats.ReconnectWait(time.Second),
		nats.DisconnectErrHandler(func(nc *nats.Conn, err error) {
			if err != nil {
				logger.Warn().Err(err).Msg("NATS disconnected")
			}
		}),
		nats.ReconnectHandler(func(nc *nats.Conn) {
			logger.Info().Msg("NATS reconnected")
		}),
	)
	if err != nil {
		return err
	}
	l.conn = conn

	// Create JetStream context
	js, err := jetstream.New(conn)
	if err != nil {
		conn.Close()
		return err
	}
	l.js = js

	// Get or wait for stream
	stream, err := l.waitForStream(ctx)
	if err != nil {
		conn.Close()
		return err
	}

	// Create consumer
	consumer, err := stream.CreateOrUpdateConsumer(ctx, jetstream.ConsumerConfig{
		Name:          ConsumerName,
		Durable:       ConsumerName,
		FilterSubject: Subject,
		DeliverPolicy: jetstream.DeliverNewPolicy,
		AckPolicy:     jetstream.AckExplicitPolicy,
		MaxDeliver:    3,
		AckWait:       30 * time.Second,
	})
	if err != nil {
		conn.Close()
		return err
	}
	l.consumer = consumer

	logger.Info().Msg("User registered listener started")

	// Start consuming
	go l.consume(ctx)

	return nil
}

func (l *UserRegisteredListener) waitForStream(ctx context.Context) (jetstream.Stream, error) {
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()

	timeout := time.After(2 * time.Minute)

	for {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-timeout:
			return nil, nats.ErrStreamNotFound
		case <-ticker.C:
			stream, err := l.js.Stream(ctx, StreamName)
			if err == nil {
				return stream, nil
			}
			logger.Debug().Msg("Waiting for USERS stream...")
		}
	}
}

func (l *UserRegisteredListener) consume(ctx context.Context) {
	for {
		select {
		case <-l.stopCh:
			return
		case <-ctx.Done():
			return
		default:
			msgs, err := l.consumer.Fetch(10, jetstream.FetchMaxWait(5*time.Second))
			if err != nil {
				if err != nats.ErrTimeout && err != context.DeadlineExceeded {
					logger.Error().Err(err).Msg("Error fetching messages")
				}
				time.Sleep(time.Second)
				continue
			}

			for msg := range msgs.Messages() {
				l.handleMessage(ctx, msg)
			}
		}
	}
}

func (l *UserRegisteredListener) handleMessage(ctx context.Context, msg jetstream.Msg) {
	var event UserRegisteredEvent
	if err := json.Unmarshal(msg.Data(), &event); err != nil {
		logger.Error().Err(err).Msg("Failed to unmarshal event")
		msg.Nak()
		return
	}

	logger.Info().
		Str("userId", event.UserID).
		Str("email", event.Email).
		Msg("Processing user.registered event")

	// Create profile
	_, err := l.profileService.CreateInitialProfile(ctx, event.UserID, event.FullName, event.Email)
	if err != nil {
		logger.Error().Err(err).
			Str("userId", event.UserID).
			Msg("Failed to create profile")
		msg.Nak()
		return
	}

	logger.Info().
		Str("userId", event.UserID).
		Msg("Profile created successfully")

	msg.Ack()
}

func (l *UserRegisteredListener) Stop() {
	close(l.stopCh)
	if l.conn != nil {
		l.conn.Drain()
		l.conn.Close()
	}
	logger.Info().Msg("User registered listener stopped")
}

File: auth-profile-nestjs/profile/src/modules/listeners/user-registered.listener.ts
import { Injectable, Logger, OnModuleInit, OnModuleDestroy } from '@nestjs/common';
import { connect, NatsConnection, JetStreamClient, JetStreamManager, ConsumerConfig, AckPolicy, DeliverPolicy } from 'nats';
import { ProfileService } from '../profile/services/profile.service';

interface UserRegisteredEvent {
  userId: string;
  email: string;
  fullName: string;
  registeredAt: string;
}

@Injectable()
export class UserRegisteredListener implements OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(UserRegisteredListener.name);
  private natsConnection: NatsConnection | null = null;
  private isRunning = false;

  constructor(private readonly profileService: ProfileService) {}

  async onModuleInit() {
    try {
      await this.start();
    } catch (error) {
      this.logger.warn('Failed to start user registered listener', error);
    }
  }

  async onModuleDestroy() {
    await this.stop();
  }

  private async start(): Promise<void> {
    const natsUrl = process.env.NATS_URL || 'nats://localhost:4222';

    try {
      this.natsConnection = await connect({ servers: natsUrl });
      const js: JetStreamClient = this.natsConnection.jetstream();
      const jsm: JetStreamManager = await this.natsConnection.jetstreamManager();

      // Ensure stream exists
      try {
        await jsm.streams.info('USERS');
      } catch {
        // Stream will be created by auth service, wait for it
        this.logger.warn('USERS stream not found, waiting for auth service to create it');
        return;
      }

      // Create durable consumer
      const consumerConfig: Partial<ConsumerConfig> = {
        durable_name: 'profile-service-user-consumer',
        ack_policy: AckPolicy.Explicit,
        deliver_policy: DeliverPolicy.New,
        filter_subject: 'user.registered',
      };

      try {
        await jsm.consumers.add('USERS', consumerConfig);
      } catch {
        // Consumer might already exist
      }

      // Start consuming
      this.isRunning = true;
      this.consume(js);

      this.logger.log('User registered listener started');
    } catch (error) {
      this.logger.error('Failed to start user registered listener', error);
      throw error;
    }
  }

  private async consume(js: JetStreamClient): Promise<void> {
    try {
      const consumer = await js.consumers.get('USERS', 'profile-service-user-consumer');

      while (this.isRunning) {
        try {
          const messages = await consumer.fetch({ max_messages: 10, expires: 5000 });

          for await (const msg of messages) {
            try {
              const event: UserRegisteredEvent = JSON.parse(new TextDecoder().decode(msg.data));
              await this.handleUserRegistered(event);
              msg.ack();
            } catch (error) {
              this.logger.error('Failed to process message', error);
              msg.nak();
            }
          }
        } catch (error) {
          if (this.isRunning) {
            this.logger.error('Error fetching messages', error);
            await this.sleep(1000);
          }
        }
      }
    } catch (error) {
      this.logger.error('Consumer error', error);
    }
  }

  private async handleUserRegistered(event: UserRegisteredEvent): Promise<void> {
    this.logger.log(`Creating profile for user: ${event.userId}`);

    try {
      // Create initial profile with user info from auth service
      const profile = await this.profileService.createInitialProfile(
        event.userId,
        event.fullName,
        event.email,
      );

      this.logger.log(`Profile created for user: ${event.userId}, display name: ${profile.displayName}`);
    } catch (error) {
      this.logger.error(`Failed to create profile for user: ${event.userId}`, error);
      throw error; // Will cause NAK and retry
    }
  }

  private async stop(): Promise<void> {
    this.isRunning = false;
    if (this.natsConnection) {
      await this.natsConnection.drain();
      this.natsConnection = null;
      this.logger.log('User registered listener stopped');
    }
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}
